Streamline the Docker build process

feuGeneA opened the pull request.

feuGeneA committed:
Stop re-cloning the git repo for every image build

Stop re-cloning the repo while building the docker image. If you're
using this Dockerfile then you've already got the repo checked out
locally, so just COPY those files into the image rather than initiating
another network request.

I don't fully understand why we need the other changes in this commit
(the user and group permissions management, etc), but it works!, and
I've verified that it fails to work if I cut out any of those other
pieces.  There's potential for more cleanup.

feuGeneA committed:
Explain how to extract and re-use cache from host

feuGeneA committed:
Build tool env independent of codebase changes

Avoid copying the entire codebase into the image before building the
tool environment, so that actual code changes (outside of eg
package.json for dependency updates, or .git modifications for `dapp
update`) won't trigger rebuilds of the entire tool environment.

feuGeneA committed:
follow .env convention, cut docker build arg

Stop making a special case of the RPC URL just for the Docker
environment.  Now that we're using the host's git working tree, their
regular old .env file can be expected to be present, so quit
paramaterizing it.

feuGeneA committed:
Add missing `-t` argument to docker examples

mds1 commented on Dockerfile:
```diff
@@ -4,13 +4,22 @@
 #
-#   docker build --build-arg ETH_RPC_URL=... -t sim .
-#   docker run sim benchmark-hardhat
+#   docker build --build-arg -t sim .
```
Was this intended to be changed? Running `docker build --build-arg -t sim .` gave the below error, but the original command still worked

```
"docker build" requires exactly 1 argument.
See 'docker build --help'.

Usage:  docker build [OPTIONS] PATH | URL | -

Build an image from a Dockerfile
```

Also, where does the RPC URL come from if we no longer pass it in a build arg?

feuGeneA replied:
Ugh. I'm sorry I haven't done a better job of verifying these example commands.  I've now added a simple test script for this Dockerfile, in https://github.com/mds1/convex-shutdown-simulation/pull/8/commits/ef583aedd72191bdd70021e38e04b477e4f3226d . It would be nice if a CI system could run it one day :smile: 

Yes, I intended to remove that `--build-arg` argument, and I did it in https://github.com/mds1/convex-shutdown-simulation/pull/8/commits/471235fe91784349a0e1fb009d54f9841ace133d .  `--build-arg` should be followed by a `KEY=VALUE` pair, and without `ETH_RPC_URL=...` it's nonsensical.

The original command still worked because you're always free to supply extra build arguments, whether the Dockerfile actually refers to them or not.

RPC URL comes from your `.env` file, just like for the non-Docker use cases. Now that we're building the image from the local git Working Tree (as of https://github.com/mds1/convex-shutdown-simulation/pull/8/commits/89046279ee0f30be507351ad9ae57ac77b5f3a2c) the local `.env` file will get copied into the image when it's built, and it will be found at run time just like when running outside of a container.

feuGeneA committed:
Remove erroneous `--build-arg` from example

feuGeneA committed:
Add test to verify example commands in Dockerfile

feuGeneA committed:
Utilize and explain tweaking 1 layer per benchmark

feuGeneA committed:
Add executable bit to testDockerfile.sh

feuGeneA commented on testDockerfile.sh:
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+EXAMPLE_COMMANDS=$(grep -e "^#   " Dockerfile | sed -e "s/^#   //"|head -2)
```
Oops! This was a relic from my local testing!
```suggestion
EXAMPLE_COMMANDS=$(grep -e "^#   " Dockerfile | sed -e "s/^#   //")
```

feuGeneA committed:
Remove oops from testDockerfile.sh

Addresses review comment https://github.com/mds1/convex-shutdown-simulation/pull/8/files#r796621496

feuGeneA committed:
exclude some files from the Image build

These files aren't needed within the image, but prior to this change,
any updates to these files would trigger a rebuild of the `COPY . .`
layer, which is wasteful. Now, such changes won't trigger such rebuilds.

feuGeneA committed:
Better document testDockerfile.sh

mds1 reviewed (approved):
LGTM and all working now! I know you just pushed some cleanup commits recently, so let me know when this is ready to merge

feuGeneA committed:
testDockerfile: use EXAMPLE_CMD_PREFIX in sed too

feuGeneA commented:
Ready to merge! Thanks!

feuGeneA committed:
Add one more example command, USING the cache

feuGeneA committed:
comments: be official: Image not image

feuGeneA committed:
comments: tiny style change

feuGeneA commented:
Still ready :)

feuGeneA committed:
Comments: tweak BEWARE warning

Now that we're using .env instead of a build arg, this comment could
benefit from a little rewording.

feuGeneA commented:
I may never stop :smile: . This is definitely ready to merge. If I come with more changes after you merge, I'll just open another PR. Thanks again.

mds1 commented:
heh alright I'll merge now!
